pkg/ipam/clusterpool_v2beta2.go

```diff
@@ -0,0 +1,474 @@
+	"github.com/cilium/cilium/pkg/k8s"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/trigger"
+)
+
+type poolPair struct {
+	v4 *podCIDRPool
+	v6 *podCIDRPool
+}
+
+type preAllocValue int
+type preAllocMap map[string]preAllocValue
+
+func parsePreAllocMap(conf map[string]string) (preAllocMap, error) {
+	m := make(map[string]preAllocValue, len(conf))
+	for pool, s := range conf {
+		value, err := strconv.ParseInt(s, 10, 32)
+		if err != nil {
+			return nil, fmt.Errorf("invalid pre-alloc value for pool %q: %w", pool, err)
+		}
+		m[pool] = preAllocValue(value)
+	}
+
+	return m, nil
+}
+
+type clusterPoolManager struct {
+	mutex *lock.Mutex
+	conf  Configuration
+	owner Owner
+
+	preallocMap  preAllocMap
+	pools        map[string]*poolPair
+	poolsUpdated chan struct{}
+
+	node *ciliumv2.CiliumNode
+
+	controller  *controller.Manager
+	k8sUpdater  *trigger.Trigger
+	nodeUpdater nodeUpdater
+
+	finishedRestore bool
+}
+
+var _ Allocator = (*clusterPoolV2Allocator)(nil)
+
+func newClusterPoolManager(conf Configuration, nodeWatcher nodeWatcher, owner Owner, clientset nodeUpdater) *clusterPoolManager {
+	preallocMap, err := parsePreAllocMap(option.Config.IPAMClusterPoolNodePreAlloc)
+	if err != nil {
+		log.WithError(err).Fatalf("Invalid %s flag value", option.IPAMClusterPoolNodePreAlloc)
+	}
+
+	k8sController := controller.NewManager()
+	k8sUpdater, err := trigger.NewTrigger(trigger.Parameters{
+		MinInterval: 15 * time.Second,
+		TriggerFunc: func(reasons []string) {
+			k8sController.TriggerController(clusterPoolStatusControllerName)
+		},
+		Name: clusterPoolStatusTriggerName,
+	})
+	if err != nil {
+		log.WithError(err).Fatal("Unable to initialize CiliumNode synchronization trigger")
+	}
+
+	c := &clusterPoolManager{
+		mutex:           &lock.Mutex{},
+		owner:           owner,
+		conf:            conf,
+		preallocMap:     preallocMap,
+		pools:           map[string]*poolPair{},
+		poolsUpdated:    make(chan struct{}, 1),
+		node:            nil,
+		controller:      k8sController,
+		k8sUpdater:      k8sUpdater,
+		nodeUpdater:     clientset,
+		finishedRestore: false,
+	}
+
+	// Subscribe to CiliumNode updates
+	nodeWatcher.RegisterCiliumNodeSubscriber(c)
+	owner.UpdateCiliumNodeResource()
+
+	c.waitForAllPools()
+
+	return c
+}
+
+func (c *clusterPoolManager) waitForAllPools() {
+	ctx := context.Background()
+	for pool := range c.preallocMap {
+		if c.conf.IPv4Enabled() {
+			c.waitForPool(ctx, IPv4, pool)
+		}
+		if c.conf.IPv6Enabled() {
+			c.waitForPool(ctx, IPv6, pool)
+		}
+	}
+}
+
+func (c *clusterPoolManager) waitForPool(ctx context.Context, family Family, poolName string) {
+	timer, stop := inctimer.New()
+	defer stop()
+	for {
+		c.mutex.Lock()
+		switch family {
+		case IPv4:
+			if p, ok := c.pools[poolName]; ok && p.v4 != nil && p.v4.hasAvailableIPs() {
```

Comment:
from meeting: what happens if the pool happened to be full on startup?

Reply:
answer: we haven't marked any existing reservations yet, so this is OK (but confusing)